File: loomc-java/src/main/java/loom/compiler/parser/objects/StatementParser.java
package loom.compiler.parser.objects;

import loom.compiler.ast.Expr;
import loom.compiler.ast.Stmt;
import loom.compiler.ast.statements.BlockStmt;
import loom.compiler.ast.statements.ExpressionStmt;
import loom.compiler.ast.statements.ReturnStmt;
import loom.compiler.diagnostics.ParseError;
import loom.compiler.parser.Parser;
import loom.compiler.token.Keywords;
import loom.compiler.token.Token;
import loom.compiler.token.TokenSpan;
import loom.compiler.token.TokenType;

public class StatementParser {

	private final Parser parser;

	public StatementParser(Parser parser) {
		this.parser = parser;
	}

	public Stmt parseStatement () {
		try {
			if (parser.match(TokenType.LBRACE)) {
				return new BlockStmt(parser.getBlockParser().parseBlock(), parser.previous().getSpan());
			}

			if (parser.match(TokenType.KEYWORD, Keywords.Keyword.RETURN)) {
				Token returnToken = parser.previous();
				Expr value = null;
				if (!parser.check(TokenType.SEMICOLON)) {
					value = parser.getPrecedenceParser().parseExpression();
				}
				Token semicolon = parser.consume(TokenType.SEMICOLON, "Expected ';' after return");
				TokenSpan span = returnToken.getSpan().merge(semicolon.getSpan());
				return new ReturnStmt(value, span);
			}

			// Control flow statements
			if (parser.match(TokenType.KEYWORD, Keywords.Keyword.IF)) {
				return parser.getIfStatementParser().parseIfStatement();
			}

			if (parser.match(TokenType.KEYWORD, Keywords.Keyword.WHILE)) {
				return parser.getWhileStatementParser().parseWhileStatement();
			}

			if (parser.match(TokenType.KEYWORD, Keywords.Keyword.FOR)) {
				return parser.getForStatementParser().parseForStatement();
			}

			if (parser.match(TokenType.KEYWORD, Keywords.Keyword.DO)) {
				return parser.getDoWhileStatementParser().parseDoWhileStatement();
			}

			// Break and continue statements
			Stmt breakContinue = parser.getBreakContinueParser().parseBreakOrContinue();
			if (breakContinue != null) {
				return breakContinue;
			}

			// Expression statement fallback
			Expr expr = parser.getPrecedenceParser().parseExpression();
			Token semi = parser.consume(TokenType.SEMICOLON, "Expected ';' after expression");
			TokenSpan span = expr.getSpan().merge(semi.getSpan());
			return new ExpressionStmt(expr, span);
		} catch (ParseError e) {
			parser.synchronize();
			return null;
		}
	}

}

File: loomc/lib/statement_parser.ml
(** Statement parser for basic statements *)

open Token_types
open Token
open Ast.Statements
open Ast.Expressions

(** Parse a return statement *)
let parse_return parser =
  match Parser.peek parser with
  | Some token when Token.token_type token = Keyword && Token.value token = "return" ->
      let start_span = Token.span token in
      let _ = Parser.advance parser in
      
      (* Parse optional return value *)
      let return_value = 
        if Parser.check parser Semicolon then None
        else (
          match Expression_parser.parse_expression parser with
          | Some expr -> Some expr
          | None -> None
        )
      in
      
      let _ = Parser.consume parser Semicolon "Expected ';' after return statement" in
      Some (Ast.Statements.return return_value start_span)
  | _ -> None

(** Parse an expression statement *)
let parse_expression_statement parser =
  match Expression_parser.parse_expression parser with
  | Some expr ->
      let _ = Parser.consume parser Semicolon "Expected ';' after expression" in
      Some (Ast.Statements.expression expr (Ast.Expressions.span expr))
  | None -> None

(** Parse a block statement *)
let parse_block parser =
  match Parser.peek parser with
  | Some token when Token.token_type token = Lbrace ->
      let start_span = Token.span token in
      let _ = Parser.advance parser in
      
      let rec parse_statements acc =
        if Parser.is_at_end parser then List.rev acc
        else (
          match Parser.peek parser with
          | Some token when Token.token_type token = Rbrace ->
              let _ = Parser.advance parser in
              List.rev acc
          | _ ->
              (match parse_statement parser with
               | Some stmt -> parse_statements (stmt :: acc)
               | None -> parse_statements acc)
        )
      in
      
      let statements = parse_statements [] in
      Some (Ast.Statements.block statements start_span)
  | _ -> None

(** Parse a single statement *)
let parse_statement parser =
  (* Try different statement types in order *)
  match parse_return parser with
  | Some stmt -> Some stmt
  | None ->
      (match parse_block parser with
       | Some stmt -> Some stmt
       | None ->
           (match parse_expression_statement parser with
            | Some stmt -> Some stmt
            | None -> None)) 